### DISCUSSION TITLE
plesk + fail2ban = reset counter after succesful login

### DISCUSSION THREAD

**Discussion by zandadoum** [fail2ban/fail2ban#3319] [Q&A]
Hello,

I have some trouble with my mail server (Plesk Obsidian on CentOS Linux 7.9.2009)
I don't have access to SSH to it, but I do to the config panel of course

The problem is that many of my customers who use an android device and connect to IMAP get ip banned

because a problem with plesk+android+digest_md5

what happens is that a android device connects with one auth method, fails, then tries another method and succeeds

problem is: as the failed counter is not reset once auth succedeed, when the android device tries next time (15min for example) this all happens again and they end up banned after a few tries

obvisouly the best would be to solve the problem at it's root, but as i can't depend on android updates, plesk updates and just removing digest_md5 from the config breaks outlook users and gets added again with every plesk upgrade....

what i would like is the fail counter to reset after a success.
it seems this might have been implemented into fail2ban some time ago, or at least it has to a certain degree:
https://github.com/fail2ban/fail2ban/issues/2102

but i have no idea how to exactly use that.

here's my dovecot jail config, can someone tell me if it's possible?

```
[plesk-dovecot]
enabled = true
filter = plesk-dovecot
action = iptables-multiport[name="plesk-dovecot", port="imap,imaps,pop3,pop3s,4190"]
logpath = /var/log/maillog
maxretry = 3
```

thanks in advance

**Comment by sebres** [comment]
> the failed counter is not reset once auth succeeded

There is a small issue with such reset-approach: some evildoer (which also has a legitimate user account) can try to login as another user and after few failed attempts login with his original account even to reset counter. So it can make a brute force for other user accounts without triggering a ban.

To solve that fail2ban must distinguish user names, etc.

Anyway to help you, we would need an excerpt from the log with failed and succeeded attempts.

**Reply by zandadoum** [reply]
hello.

i am not concerned with "some evildoer" as i have a pretty tight control over who has accounts on this server and i get notifications when "something weird is afoot"

as for the logs, as mentioned i dont have access to SSH, so all i can see is what plesk control panel show me from /var/log/fail2ban.log and that's really limited, only sais "IP found, IP banned" and such short one-liners

i can however see mail logs on a more detailed way and this is what i see:

```
2022-07-15 14:01:24	dovecot	service=imap, user=<USER_EMAIL>, ip=[<USER_IP>]. Disconnected: Connection closed (SELECT finished 0.158 secs ago) rcvd=179, sent=1290
2022-07-15 14:01:15	dovecot	imap-login: Login: user=<<USER_EMAIL>>, method=PLAIN, rip=<USER_IP>, lip=<SERVER_IP>, mpid=30227, TLS, session=<wl44xtbj3thTMlQj>
2022-07-15 14:01:11	dovecot	imap-login: Disconnected: Connection closed (auth failed, 1 attempts in 3 secs): user=<<USER_EMAIL>>, method=DIGEST-MD5, rip=<USER_IP>, lip=<SERVER_IP>, TLS, session=<QFcCxtbj2thTMlQj>
```
as you see (starting at bottom) it tries to get in with digest-md5 and fails
then gets through properly
then disconnects properly

while trying to find a solution for this, i came up with the article i linked in my first message. they mentioned it can be done with `<F-MLFGAINED>` but i don't know how to implement this.

**Reply by sebres** [reply]
> i can however see mail logs on a more detailed way and this is what i see

You log excerpt is too short to me. Please provide larger excerpt, and don't blacken the data this way.... Just use 192.0.2.123 for user-IP, 127.0.0.1 for server IP, and some-name@example.com for mail.

I am unsure `<F-MLFGAINED>` would help here, because normally it will be used for some a bit longer "session" (with several failed attempts and 1 succeeded attempt or disconnect), but lets wait for larger excerpt.
Another variant would be #1243, which is not yet fulfilled (due to reason I mentioned above with different user-accounts) and therefore not released.

**Reply by zandadoum** [reply]
the rest of the log is exactly the same 3 lines at different hours of the day, every 15minutes (which is when his phone checks email).
no real need to delve into that any further
and #1243 has been opened since 2015, not receive any significant updates since 2020 and all related comments from 2y ago have been closed. i highly doubt that is ever gonna see the light.

so, i guess it can't be done and it won't be done.

**Reply by sebres** [reply]
> and #1243 has been opened since 2015, ... i highly doubt that is ever gonna see the light.

Well, I can try to rebase it to latest versions and check what we need to fulfill it... just if it gets released at some point, you'd probably anyway don't see it soon on your system due to certain distribution policies (and I guess a manual upgrade is impossible).

As regards the log, if it really contains only this 3 lines, `<F-MLFGAINED>` can not be an option, because despite to reset a failure it would be enough to consider successful login message, but to notice a failure it'd expect still one message of another type (which you seem to don't have).
However, if you don't support DIGEST-MD5 at all, what you could do (unless #1243 gets released) would be new regex covering only DIGEST-MD5 case, so one would ignore them completely (sure it could cause a lot of spam failures in log if someone trying a brute-force using DIGEST-MD5, but it is suitable workaround if they have no chance at all by that method).
```ini
[dovecot]
enabled = true
...
# ignore DIGEST-MD5 method to avoid mistaken bans of legitimate users with strange clients, trying DIGEST-MD5 before PLAIN:
failregex = ^(?:Aborted login|Disconnected|Remote closed connection|Client has quit the connection)(?:: (?:[^\(]+|\w+\([^\)]*\))+)? \((?:auth failed, \d+ attempts(?: in \d+ secs)?|tried to use (?:disabled|disallowed) \S+ auth|proxy dest auth failed)\):(?: user=<<F-USER>[^>]*</F-USER>>,)?(?: method=(?:<F-NOFAIL>DIGEST-MD5</F-NOFAIL>|\S+),)? rip=<HOST>(?:[^>]*(?:, session=<\S+>)?)\s*$
            %(known/failregex)s
```
Although I assume you're trying to repair it from wrong end, because I don't think it would be really large issue to implement your authentication (SQL?) to support DIGEST-MD5 (it must just compare the MD5-hash von "user:realm:password" (or whatever other nonce) instead of the password directly:
```diff
- WHERE tab.user = ? and tab.pwd = ?
+ WHERE tab.user = ? and md5(concat(tab.user,':some-dovecot-realm:',tab.pwd)) = ?
```
It is also a bit safer (if you rewrite the auth-mechanism correctly), because you don't need to store the plain passwords in database (only the hashes), so if database gets stolen (by some hacking attempt), it'd be hard for the villains to obtain the real passwords (expects very long time to sort through the hashes brute-forcing them).
